File: solutions/039.combination-sum/combination-sum.java
class Solution {
    public
        ArrayList<Integer> item = new ArrayList<Integer>();
        List<List<Integer>> res = new ArrayList<>();
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        int length = candidates.length;
        Arrays.sort(candidates);
        if(length==0){
            return res;
        }
        helper(candidates, target, 0);
        return res;
    }
    
    public void helper(int[] nums, int target, int index){
        for(int i =index;i<nums.length;i++){
            if(target<nums[i]){
                break;
            }
            if(target == nums[i]){
                item.add(nums[i]);
                res.add(new ArrayList<Integer>(item));
                item.remove(item.size()-1);
                break;
            }
            if(target>nums[i]){
                item.add(nums[i]);
                helper(nums, target-nums[i], i);
                item.remove(item.size()-1);
            }
        }
    }
}

File: solutions/039.combination-sum/combination-sum.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function(candidates, target) {
    const res = [];
    helper(candidates, 0, [], 0, target);
    return res;
    
    function helper(nums, index, arr, sum, target){
        if(sum === target){
            res.push([...arr]);
            return;
        }
        if(sum > target || index >= nums.length ){
            return;
        }
        
        for(let i = index; i < nums.length; i++){
            arr.push(nums[i]);
            helper(nums, i, arr, sum + nums[i], target);
            arr.pop();
        }
    }
    
};